x-pack/plugin/wildcard/src/main/java/org/elasticsearch/xpack/wildcard/mapper/NGramAutomaton.java

```diff
@@ -0,0 +1,374 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+package org.elasticsearch.xpack.wildcard.mapper;
+
+import org.apache.lucene.util.automaton.Automaton;
+import org.apache.lucene.util.automaton.Transition;
+import org.elasticsearch.xpack.wildcard.mapper.regex.And;
+import org.elasticsearch.xpack.wildcard.mapper.regex.AutomatonTooComplexException;
+import org.elasticsearch.xpack.wildcard.mapper.regex.Expression;
+import org.elasticsearch.xpack.wildcard.mapper.regex.ExpressionSource;
+import org.elasticsearch.xpack.wildcard.mapper.regex.False;
+import org.elasticsearch.xpack.wildcard.mapper.regex.Leaf;
+import org.elasticsearch.xpack.wildcard.mapper.regex.Or;
+import org.elasticsearch.xpack.wildcard.mapper.regex.True;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * A finite automaton who's transitions are ngrams that must be in the string or
+ * ngrams we can't check for. Not thread safe one bit.
+ */
+public class NGramAutomaton {
+    private final Automaton source;
+    private final int gramSize;
+    private final int maxExpand;
+    private final int maxStatesTraced;
+    private final int maxTransitions;
+    private final List<NGramState> initialStates = new ArrayList<>();
+    private final List<NGramState> acceptStates = new ArrayList<>();
+    private final Map<NGramState, NGramState> states = new HashMap<>();
+
+    /**
+     * Build it.
+     * @param source automaton to convert into an ngram automaton
+     * @param gramSize size of the grams to extract
+     * @param maxExpand Maximum size of range transitions to expand into single
+     *            transitions. Its roughly analogous to the number of character
+     *            in a character class before it is considered a wildcard for
+     *            optimization purposes.
+     * @param maxStatesTraced maximum number of states traced during automaton
+     *            functions. Higher number allow more complex automata to be
+     *            converted to ngram expressions at the cost of more time.
+     */
+    public NGramAutomaton(Automaton source, int gramSize, int maxExpand, int maxStatesTraced, int maxTransitions) {
+        this.source = source;
+        this.gramSize = gramSize;
+        this.maxExpand = maxExpand;
+        this.maxStatesTraced = maxStatesTraced;
+        this.maxTransitions = maxTransitions;
+        if (source.getNumStates() == 0) {
+            return;
+        }
+        // Build the initial states using the first gramSize transitions
+        int[] codePoints = new int[gramSize - 1];
+        buildInitial(codePoints, 0, 0);
```

Comment:
I know this is probably a mistake that I made years ago, but `buildInitial` looks to return a boolean that we throw away.